### PR TITLE
Ports: Add perl5

### PR DESCRIPTION
**Pull request by tarob0ba**
The moment we've all been craving. This ports over `perl` v5.38.0 (latest stable), which opens the doors to future opportunities.
This also would address #1613.

The patches included are mostly self explanatory.
If you have an ARM Mac, I'd love to hear whether or not this works (as I cannot test it on my end).

Screenshot:
Screenshot of `perl -v` and a "Hello, world!" script

### CODE REVIEW

**Review by tcl3 — COMMENTED** [review]
It works well based on my very limited testing, thanks for getting this working. There's just a few small things I noticed.

File: Ports/perl5/package.sh
Comment (tarob0ba) [reply Ports/perl5/package.sh]: thanks! i've taken it out.

File: Ports/perl5/package.sh
Comment (tarob0ba) [reply Ports/perl5/package.sh]: whoops! i completely missed that. we're now making use of `configopts`.

File: Ports/perl5/package.sh
Comment (tcl3) [inline Ports/perl5/package.sh:49]: You should be able to remove this entirely.

**Review by tcl3 — APPROVED** [review]
I can confirm that the most important use case for perl works correctly!

cowsay

File: Ports/perl5/package.sh
Comment (EWouters) [inline Ports/perl5/package.sh:19]:
```suggestion
    run cat <<- 'EOH' > "$PWD/perl-${version}/cnf/hints/serenity"
```

File: Ports/perl5/package.sh
Comment (EWouters) [inline Ports/perl5/package.sh:26]: These lines are in reverse order.
```suggestion
    run chmod +w "${configscript}"
    run cp -r "$PWD/perl-cross-1.5/"* "$PWD/perl-${version}/"
```

File: Ports/perl5/package.sh
Comment (EWouters) [inline Ports/perl5/package.sh:18]: This is more readable I think.
```suggestion
    makeopts+=("--ignore-errors")
```

**Review by EWouters — CHANGES_REQUESTED** [review]
[This patch](https://discord.com/channels/830522505605283862/830807158047244329/1134179041692352542) is required for the the Clang build as well.

**Comment by tarob0ba** [comment]
I’ll put that in.

File: Ports/perl5/package.sh
Comment (EWouters) [inline Ports/perl5/package.sh:19]: Other ports don't usually do this. There may be a canonical solution for this.

Comment (tarob0ba) [reply Ports/perl5/package.sh:19]: Perhaps a FIXME would be adequate-- I'll wait for a maintainer to answer.

**Comment by tarob0ba** [comment]
Undrafted as macOS build now works, assuming GNU sed and binutils are installed (`brew install binutils gnu-sed`).

File: Ports/perl5/package.sh
Comment (gmta) [reply Ports/perl5/package.sh:19]: For now, ports assume all required host binaries are available on `$PATH`. This is not ideal but it has worked for now. Feel free to leave in a `FIXME` but it's not specific to your port, it's the same for all ports that require at least one other host dependency. It might be better to open an issue for this to discuss possible approaches?

**Comment by gmta** [comment]
Building on Linux, I almost immediately get the following error:

```
[perl5/fetch] Expecting sha256sum: eca551caec3bc549a4e590c0015003790bdd1a604ffe19cc78ee631d51f7072e
[perl5/fetch] sha256sum(perl5-5.38.0.tar.xz) = 'eca551caec3bc549a4e590c0015003790bdd1a604ffe19cc78ee631d51f7072e'
[perl5/fetch] Expecting sha256sum: d744a390939e2ebb9a12f6725b4d9c19255a141d90031eff90ea183fdfcbf211
[perl5/fetch] sha256sum(perl-cross.1.5.tar.gz) = 'd744a390939e2ebb9a12f6725b4d9c19255a141d90031eff90ea183fdfcbf211'
[perl5/fetch] + tar -xf /path/to/serenity/Ports/perl5/perl5-5.38.0.tar.xz (nocd)
[perl5/fetch] + touch .perl5-5.38.0.tar.xz_extracted
[perl5/fetch] + tar -xf /path/to/serenity/Ports/perl5/perl-cross.1.5.tar.gz (nocd)
[perl5/fetch] + touch .perl-cross.1.5.tar.gz_extracted
[perl5/fetch] + chmod +w configure
[perl5/fetch] chmod: cannot access 'configure': No such file or directory
Error in step perl5/fetch (status=1)
```

File: Ports/perl5/package.sh
Comment (tarob0ba) [reply Ports/perl5/package.sh:19]: I've opened #20238.

**Comment by tarob0ba** [comment]
Should be fixed, and build should now be working properly on macOS and Linux.

**Comment by tarob0ba** [comment]
Should be fixed now, and build now seems to work on macOS and Linux again.

File: Ports/perl5/package.sh
Comment (EWouters) [inline Ports/perl5/package.sh:35]:
```suggestion
    if [ "$(uname -s)" = "Darwin" ]; then
        cat <<- 'EOH' > "$PWD/perl-${version}/readelf"
#!/bin/bash
exec gobjdump "$@"
EOH
        run chmod +x "$PWD/perl-${version}/readelf"
    fi
```

**Review by EWouters — COMMENTED** [review]
It builds on macOS with both toolchains now :^)

**Comment by EWouters** [comment]
> Building on Linux, I almost immediately get the following error:
> 
> ```
> [perl5/fetch] Expecting sha256sum: eca551caec3bc549a4e590c0015003790bdd1a604ffe19cc78ee631d51f7072e
> [perl5/fetch] sha256sum(perl5-5.38.0.tar.xz) = 'eca551caec3bc549a4e590c0015003790bdd1a604ffe19cc78ee631d51f7072e'
> [perl5/fetch] Expecting sha256sum: d744a390939e2ebb9a12f6725b4d9c19255a141d90031eff90ea183fdfcbf211
> [perl5/fetch] sha256sum(perl-cross.1.5.tar.gz) = 'd744a390939e2ebb9a12f6725b4d9c19255a141d90031eff90ea183fdfcbf211'
> [perl5/fetch] + tar -xf /path/to/serenity/Ports/perl5/perl5-5.38.0.tar.xz (nocd)
> [perl5/fetch] + touch .perl5-5.38.0.tar.xz_extracted
> [perl5/fetch] + tar -xf /path/to/serenity/Ports/perl5/perl-cross.1.5.tar.gz (nocd)
> [perl5/fetch] + touch .perl-cross.1.5.tar.gz_extracted
> [perl5/fetch] + chmod +w configure
> [perl5/fetch] chmod: cannot access 'configure': No such file or directory
> Error in step perl5/fetch (status=1)
> ```

This is something related to differences in handling capital letters on macOS and Linux I think.

**Review by EWouters — APPROVED** [review]

File: Ports/perl5/package.sh
Comment (gmta) [inline Ports/perl5/package.sh:8]: Please use single quotes for strings without interpolation

File: Ports/perl5/package.sh
Comment (gmta) [inline Ports/perl5/package.sh:17]: Same here - `'Darwin'`
And the line below `'--ignore-errors'`

File: Ports/perl5/package.sh
Comment (gmta) [inline Ports/perl5/package.sh:22]: Please use one style of string interpolation through the script. I.e. `${PATH}` here and `${PWD}` below

**Comment by gmta** [comment]
It now builds for me, but it installs `perl` into `/usr/bin`. We install almost every port into `/usr/local/bin`, `/usr/local/lib` etc.

**Comment by tarob0ba** [comment]
Everything you brought up should be addressed now. It turns out that I forgot to add `--prefix` :facepalm:.

**Review by ADKaster — APPROVED** [review]

**Comment by Leont** [comment]
I've made parts of this a PR to include into perl itself https://github.com/Perl/perl5/pull/21327, so that in 5.40 you don't have to patch as much (at least for native build, perl-cross is somebody else's problem). Please do send us any updates to the hints file.

**Comment by Leont** [comment]
> I've made parts of this a PR to include into perl itself [Perl/perl5#21327](https://github.com/Perl/perl5/pull/21327), so that in 5.40 you don't have to patch as much (at least for native build, perl-cross is somebody else's problem). Please do send us any updates to the hints file.

Possibly I misunderstood and you always use perl-cross?

**Comment by tarob0ba** [comment]
We always use perl-cross at the moment (as Serenity's ports are almost always built on the host), but your PR will allow us to build Perl on Serenity without most of these patches.

Thank you! :^)

**Pull request merged**
